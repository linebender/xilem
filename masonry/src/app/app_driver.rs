// Copyright 2024 the Xilem Authors
// SPDX-License-Identifier: Apache-2.0

use crate::event_loop_runner::MasonryState;
use crate::{Action, RenderRoot, WidgetId};

/// Context for the [`AppDriver`] trait.
///
/// Currently holds a reference to the [`RenderRoot`].
pub struct DriverCtx<'a> {
    // We make no guarantees about the fields of this struct, but
    // they must all be public so that the type can be constructed
    // externally.
    // This is needed for external users, whilst our external API
    // is not yet designed.
    #[doc(hidden)]
    pub render_root: &'a mut RenderRoot,
}

/// A trait for defining how your app interacts with the Masonry widget tree.
///
/// When launching your app with [`crate::event_loop_runner::run`], you need to provide
/// a type that implements this trait.
pub trait AppDriver {
    /// A hook which will be executed when a widget emits an [`Action`].
    fn on_action(&mut self, ctx: &mut DriverCtx<'_>, widget_id: WidgetId, action: Action);

    #[allow(unused_variables)]
    // reason: otherwise `state` would need to be named `_state` which behaves badly when using rust-analyzer to implement the trait
    /// A hook which will be executed when the application starts, to allow initial configuration of the `MasonryState`.
    ///
    /// Use cases include loading fonts.
    fn on_start(&mut self, state: &mut MasonryState) {}
}

impl DriverCtx<'_> {
    // TODO - Add method to create timer

    /// Access the [`RenderRoot`].
    pub fn render_root(&mut self) -> &mut RenderRoot {
        self.render_root
    }

    /// Returns true if something happened that requires a rewrite pass or a re-render.
    pub fn content_changed(&self) -> bool {
        self.render_root.needs_rewrite_passes()
    }
}

#[cfg(doctest)]
/// Doctests aren't collected under `cfg(test)`; we can use `cfg(doctest)` instead
mod doctests {
    /// ```no_run
    /// use masonry::DriverCtx;
    /// let _ctx = DriverCtx {
    ///     render_root: unimplemented!()
    /// };
    /// ```
    const _DRIVER_CTX_EXTERNALLY_CONSTRUCTIBLE: () = {};
}
