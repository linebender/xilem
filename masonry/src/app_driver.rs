// Copyright 2024 the Xilem Authors
// SPDX-License-Identifier: Apache-2.0

use crate::widget::WidgetMut;
use crate::{Action, Widget, WidgetId};

// xilem::App will implement AppDriver

pub struct DriverCtx<'a> {
    // TODO
    // This is exposed publicly for now to let people drive
    // masonry on their own, but this is not expected to be
    // stable or even supported. This is for short term
    // expedience only while better solutions are devised.
    #[doc(hidden)]
    pub main_root_widget: WidgetMut<'a, Box<dyn Widget>>,
}

pub trait AppDriver {
    fn on_action(&mut self, ctx: &mut DriverCtx<'_>, widget_id: WidgetId, action: Action);
}

impl<'a> DriverCtx<'a> {
    /// Return a [`WidgetMut`] to the root widget.
    pub fn get_root<W: Widget>(&mut self) -> WidgetMut<'_, W> {
        self.main_root_widget.downcast()
    }
}
