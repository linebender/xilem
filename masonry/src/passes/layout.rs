// Copyright 2024 the Xilem Authors
// SPDX-License-Identifier: Apache-2.0

//! The layout pass, which provides the size and position of each widget
//! before any translations applied in [`compose`](crate::passes::compose).
//! Most of the logic for this pass happens in [`Widget::layout`] implementations.

use dpi::LogicalSize;
use smallvec::SmallVec;
use tracing::{debug, info_span, trace};
use vello::kurbo::{Point, Rect, Size};

use crate::passes::recurse_on_children;
use crate::render_root::{RenderRoot, RenderRootSignal, WindowSizePolicy};
use crate::widget::WidgetState;
use crate::{BoxConstraints, LayoutCtx, Widget, WidgetPod};

// --- MARK: RUN LAYOUT ---
/// Run [`Widget::layout`] method on the widget contained in `pod`.
/// This will be called by [`LayoutCtx::run_layout`], which is itself called in the parent widget's `layout`.
pub(crate) fn run_layout_on<W: Widget>(
    parent_ctx: &mut LayoutCtx<'_>,
    pod: &mut WidgetPod<W>,
    bc: &BoxConstraints,
) -> Size {
    let id = pod.id().to_raw();
    let mut widget = parent_ctx.widget_children.get_child_mut(id).unwrap();
    let mut state = parent_ctx.widget_state_children.get_child_mut(id).unwrap();

    let _span = widget.item.make_trace_span().entered();

    let mut children_ids = SmallVec::new();
    if cfg!(debug_assertions) {
        children_ids = widget.item.children_ids();

        // We forcefully set request_layout to true for all children.
        // This is used below to check that widget.layout(..) visited all of them.
        for child_id in widget.item.children_ids() {
            let child_id = child_id.to_raw();
            let child_state = state.children.get_child_mut(child_id).unwrap().item;
            if !child_state.is_stashed {
                child_state.request_layout = true;
            }
        }
    }

    // This checks reads is_explicitly_stashed instead of is_stashed because the latter may be outdated.
    // A widget's is_explicitly_stashed flag is controlled by its direct parent.
    // The parent may set this flag during layout, in which case it should avoid calling run_layout.
    // Note that, because this check exits before recursing, run_layout can only ever be
    // reached for a widget whose parent is not stashed, which means is_explicitly_stashed
    // being false is sufficient to know the widget is non-stashed.
    if state.item.is_explicitly_stashed {
        debug_panic!(
            "Error in '{}' {}: trying to compute layout of stashed widget.",
            widget.item.short_type_name(),
            pod.id(),
        );
        state.item.size = Size::ZERO;
        return Size::ZERO;
    }

    // TODO - Not everything that has been re-laid out needs to be repainted.
    state.item.needs_paint = true;
    state.item.needs_compose = true;
    state.item.needs_accessibility = true;
    state.item.request_paint = true;
    state.item.request_compose = true;
    state.item.request_accessibility = true;

    bc.debug_check(widget.item.short_type_name());
    trace!("Computing layout with constraints {:?}", bc);

    state.item.local_paint_rect = Rect::ZERO;

    // TODO - Handle more elegantly
    // We suppress need_layout and request_layout for stashed children
    // to avoid unnecessary relayouts in corner cases.
    recurse_on_children(
        pod.id(),
        widget.reborrow_mut(),
        state.children.reborrow_mut(),
        |_, state| {
            if state.item.is_stashed {
                state.item.needs_layout = false;
                state.item.request_layout = false;
            }
        },
    );

    let new_size = {
        let mut inner_ctx = LayoutCtx {
            widget_state: state.item,
            widget_state_children: state.children.reborrow_mut(),
            widget_children: widget.children,
            global_state: parent_ctx.global_state,
        };

        // TODO - If constraints are the same and request_layout isn't set,
        // skip calling layout
        inner_ctx.widget_state.request_layout = false;
        widget.item.layout(&mut inner_ctx, bc)
    };
    if state.item.request_layout {
        debug_panic!(
            "Error in '{}' {}: layout request flag was set during layout pass",
            widget.item.short_type_name(),
            pod.id(),
        );
    }
    trace!(
        "Computed layout: size={}, baseline={}, insets={:?}",
        new_size,
        state.item.baseline_offset,
        state.item.paint_insets,
    );

    state.item.needs_layout = false;
    state.item.is_expecting_place_child_call = true;

    state.item.local_paint_rect = state
        .item
        .local_paint_rect
        .union(new_size.to_rect() + state.item.paint_insets);

    #[cfg(debug_assertions)]
    {
        let name = widget.item.short_type_name();
        for child_id in widget.item.children_ids() {
            let child_id = child_id.to_raw();
            let child_state = state.children.get_child_mut(child_id).unwrap().item;

            if child_state.is_stashed {
                continue;
            }

            if child_state.request_layout {
                debug_panic!(
                    "Error in '{}' {}: LayoutCtx::run_layout() was not called with child widget '{}' {}.",
                    name,
                    pod.id(),
                    child_state.widget_name,
                    child_state.id,
                );
            }

            if child_state.is_expecting_place_child_call {
                debug_panic!(
                    "Error in '{}' {}: LayoutCtx::place_child() was not called with child widget '{}' {}.",
                    name,
                    pod.id(),
                    child_state.widget_name,
                    child_state.id,
                );
            }

            // TODO - This check might be redundant with the code updating local_paint_rect
            let child_rect = child_state.paint_rect();
            if !state.item.local_paint_rect.contains_rect(child_rect) && state.item.clip.is_none() {
                /*debug_panic!(
                    "Error in '{}' {}: paint_rect {:?} doesn't contain paint_rect {:?} of child widget '{}' {}",
                    name,
                    pod.id(),
                    state.item.local_paint_rect,
                    child_rect,
                    child_state.widget_name,
                    child_state.id,
                );*/
                // TODO: The lack of measure causes problems with this check.
                debug!("paint rect doesn't contain paint rect of child")
            }
        }

        let new_children_ids = widget.item.children_ids();
        if children_ids != new_children_ids && !state.item.children_changed {
            debug_panic!(
                "Error in '{}' {}: children changed during layout pass",
                name,
                pod.id(),
            );
        }

        if !new_size.width.is_finite() || !new_size.height.is_finite() {
            debug_panic!(
                "Error in '{}' {}: invalid size {}",
                name,
                pod.id(),
                new_size
            );
        }
    }

    // TODO - Figure out how to deal with the overflow problem, eg:
    // What happens if a widget returns a size larger than the allowed constraints?
    // Some possibilities are:
    // - Always clip: might be expensive
    // - Display it anyway: might lead to graphical bugs
    // - Panic: too harsh?
    // Also, we need to avoid spurious crashes when we initialize the app and the
    // size is (0,0)
    // See https://github.com/linebender/xilem/issues/377

    let state_mut = parent_ctx.widget_state_children.get_child_mut(id).unwrap();
    parent_ctx.widget_state.merge_up(state_mut.item);
    state_mut.item.size = new_size;
    new_size
}

// --- MARK: ROOT ---
pub(crate) fn run_layout_pass(root: &mut RenderRoot) {
    if !root.root_state().needs_layout {
        return;
    }

    let _span = info_span!("layout").entered();

    let window_size = root.get_kurbo_size();
    let bc = match root.size_policy {
        WindowSizePolicy::User => BoxConstraints::tight(window_size),
        WindowSizePolicy::Content => BoxConstraints::UNBOUNDED,
    };

    let mut dummy_state = WidgetState::synthetic(root.root.id(), root.get_kurbo_size());
    let root_state_token = root.widget_arena.widget_states.root_token_mut();
    let root_widget_token = root.widget_arena.widgets.root_token_mut();
    let mut ctx = LayoutCtx {
        global_state: &mut root.state,
        widget_state: &mut dummy_state,
        widget_state_children: root_state_token,
        widget_children: root_widget_token,
    };

    let size = run_layout_on(&mut ctx, &mut root.root, &bc);
    ctx.place_child(&mut root.root, Point::ORIGIN);

    if let WindowSizePolicy::Content = root.size_policy {
        let new_size = LogicalSize::new(size.width, size.height).to_physical(root.scale_factor);
        if root.size != new_size {
            root.size = new_size;
            root.state.emit_signal(RenderRootSignal::SetSize(new_size));
        }
    }
}
