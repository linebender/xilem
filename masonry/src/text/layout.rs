// Copyright 2018 the Xilem Authors and the Druid Authors
// SPDX-License-Identifier: Apache-2.0

//! A type for laying out, drawing, and interacting with text.

use std::rc::Rc;

use kurbo::{Affine, Line, Point, Rect, Size};
use parley::context::RangedBuilder;
use parley::fontique::{Style, Weight};
use parley::layout::{Alignment, Cursor};
use parley::style::{Brush as BrushTrait, FontFamily, FontStack, GenericFamily, StyleProperty};
use parley::{FontContext, Layout, LayoutContext};
use vello::peniko::{self, Color, Gradient};
use vello::Scene;

use super::{Link, TextStorage};

/// A component for displaying text on screen.
///
/// This is a type intended to be used by other widgets that display text.
/// It allows for the text itself as well as font and other styling information
/// to be set and modified. It wraps an inner layout object, and handles
/// invalidating and rebuilding it as required.
///
/// This object is not valid until the [`rebuild_if_needed`] method has been
/// called. You should generally do this in your widget's [`layout`] method.
/// Additionally, you should call [`needs_rebuild_after_update`]
/// as part of your widget's [`update`] method; if this returns `true`, you will need
/// to call [`rebuild_if_needed`] again, generally by scheduling another [`layout`]
/// pass.
///
/// [`layout`]: trait.Widget.html#tymethod.layout
/// [`update`]: trait.Widget.html#tymethod.update
/// [`needs_rebuild_after_update`]: #method.needs_rebuild_after_update
/// [`rebuild_if_needed`]: #method.rebuild_if_needed
///
/// TODO: Update docs to mentionParley
#[derive(Clone)]
pub struct TextLayout<T> {
    text: T,
    // TODO: Find a way to let this use borrowed data
    scale: f32,

    brush: TextBrush,
    font: FontStack<'static>,
    text_size: f32,
    weight: Weight,
    style: Style,

    alignment: Alignment,
    max_advance: Option<f32>,

    links: Rc<[(Rect, usize)]>,

    needs_layout: bool,
    needs_line_breaks: bool,
    layout: Layout<TextBrush>,
    scratch_scene: Scene,
}

/// A custom brush for `Parley`, enabling using Parley to pass-through
/// which glyphs are selected/highlighted
#[derive(Clone, Debug, PartialEq)]
pub enum TextBrush {
    Normal(peniko::Brush),
    Highlight {
        text: peniko::Brush,
        fill: peniko::Brush,
    },
}

impl BrushTrait for TextBrush {}

impl From<peniko::Brush> for TextBrush {
    fn from(value: peniko::Brush) -> Self {
        Self::Normal(value)
    }
}

impl From<Gradient> for TextBrush {
    fn from(value: Gradient) -> Self {
        Self::Normal(value.into())
    }
}

impl From<Color> for TextBrush {
    fn from(value: Color) -> Self {
        Self::Normal(value.into())
    }
}

// Parley requires their Brush implementations to implement Default
impl Default for TextBrush {
    fn default() -> Self {
        Self::Normal(Default::default())
    }
}

/// Metrics describing the layout text.
#[derive(Debug, Clone, Copy, Default)]
pub struct LayoutMetrics {
    /// The nominal size of the layout.
    pub size: Size,
    /// The distance from the nominal top of the layout to the first baseline.
    pub first_baseline: f32,
    /// The width of the layout, inclusive of trailing whitespace.
    pub trailing_whitespace_width: f32,
    //TODO: add inking_rect
}

impl<T> TextLayout<T> {
    /// Create a new `TextLayout` object.
    pub fn new(text: T, text_size: f32) -> Self {
        TextLayout {
            text,
            scale: 1.0,

            brush: crate::theme::TEXT_COLOR.into(),
            font: FontStack::Single(FontFamily::Generic(GenericFamily::SansSerif)),
            text_size,
            weight: Weight::NORMAL,
            style: Style::Normal,

            max_advance: None,
            alignment: Default::default(),

            links: Rc::new([]),

            needs_layout: true,
            needs_line_breaks: true,
            layout: Layout::new(),
            scratch_scene: Scene::new(),
        }
    }

    /// Mark that the inner layout needs to be updated.
    ///
    /// This should be used if your `T` has interior mutability
    pub fn invalidate(&mut self) {
        self.needs_layout = true;
        self.needs_line_breaks = true;
    }

    /// Set the scaling factor
    pub fn set_scale(&mut self, scale: f32) {
        if scale != self.scale {
            self.scale = scale;
            self.invalidate();
        }
    }

    /// Set the default brush used for the layout.
    ///
    /// This is the non-layout impacting styling (primarily colour)
    /// used when displaying the text
    #[doc(alias = "set_color")]
    pub fn set_brush(&mut self, brush: impl Into<TextBrush>) {
        let brush = brush.into();
        if brush != self.brush {
            self.brush = brush;
            self.invalidate();
        }
    }

    /// Set the default font stack.
    pub fn set_font(&mut self, font: FontStack<'static>) {
        if font != self.font {
            self.font = font;
            self.invalidate();
        }
    }

    /// Set the font size.
    #[doc(alias = "set_font_size")]
    pub fn set_text_size(&mut self, size: f32) {
        if size != self.text_size {
            self.text_size = size;
            self.invalidate();
        }
    }

    /// Set the font weight.
    pub fn set_weight(&mut self, weight: Weight) {
        if weight != self.weight {
            self.weight = weight;
            self.invalidate();
        }
    }

    /// Set the font style.
    pub fn set_style(&mut self, style: Style) {
        if style != self.style {
            self.style = style;
            self.invalidate();
        }
    }

    /// Set the [`Alignment`] for this layout.
    pub fn set_text_alignment(&mut self, alignment: Alignment) {
        if self.alignment != alignment {
            self.alignment = alignment;
            self.invalidate();
        }
    }

    /// Set the width at which to wrap words.
    ///
    /// You may pass `None` to disable word wrapping
    /// (the default behaviour).
    pub fn set_max_advance(&mut self, max_advance: Option<f32>) {
        let max_advance = max_advance.map(|it| it.max(0.0));
        if self.max_advance.is_some() != max_advance.is_some()
            || self
                .max_advance
                .zip(max_advance)
                // 1e-4 is an arbitrary small-enough value that we don't care to rewrap
                .map(|(old, new)| (old - new).abs() >= 1e-4)
                .unwrap_or(false)
        {
            self.max_advance = max_advance;
            self.needs_line_breaks = true;
        }
    }

    /// Returns `true` if this layout needs to be rebuilt.
    ///
    /// This happens (for instance) after style attributes are modified.
    ///
    /// This does not account for things like the text changing, handling that
    /// is the responsibility of the user.
    pub fn needs_rebuild(&self) -> bool {
        self.needs_layout || self.needs_line_breaks
    }

    // TODO: What are the valid use cases for this, where we shouldn't use a run-specific check instead?
    // /// Returns `true` if this layout's text appears to be right-to-left.
    // ///
    // /// See [`piet::util::first_strong_rtl`] for more information.
    // ///
    // /// [`piet::util::first_strong_rtl`]: crate::piet::util::first_strong_rtl
    // pub fn text_is_rtl(&self) -> bool {
    //     self.text_is_rtl
    // }
}

impl<T: TextStorage> TextLayout<T> {
    #[track_caller]
    fn assert_rebuilt(&self, method: &str) {
        if self.needs_layout || self.needs_line_breaks {
            debug_panic!(
                "TextLayout::{method} called without rebuilding layout object. Text was '{}'",
                self.text.as_str().chars().take(250).collect::<String>()
            );
        }
    }

    /// Set the text to display.
    pub fn set_text(&mut self, text: T) {
        if !self.text.maybe_eq(&text) {
            self.text = text;
            self.invalidate();
        }
    }

    /// Returns the [`TextStorage`] backing this layout, if it exists.
    pub fn text(&self) -> &T {
        &self.text
    }

    /// Returns the [`TextStorage`] backing this layout, if it exists.
    ///
    /// Invalidates the layout and so should only be used when definitely applying an edit
    pub fn text_mut(&mut self) -> &mut T {
        self.invalidate();
        &mut self.text
    }

    /// Returns the inner Parley [`Layout`] value.
    pub fn layout(&self) -> &Layout<TextBrush> {
        self.assert_rebuilt("layout");
        &self.layout
    }

    /// The size of the laid-out text, excluding any trailing whitespace.
    ///
    /// This is not meaningful until [`Self::rebuild`] has been called.
    pub fn size(&self) -> Size {
        self.assert_rebuilt("size");
        Size::new(self.layout.width().into(), self.layout.height().into())
    }

    /// The size of the laid-out text, including any trailing whitespace.
    ///
    /// This is not meaningful until [`Self::rebuild`] has been called.
    pub fn full_size(&self) -> Size {
        self.assert_rebuilt("full_size");
        Size::new(self.layout.full_width().into(), self.layout.height().into())
    }

    /// Return the text's [`LayoutMetrics`].
    ///
    /// This is not meaningful until [`Self::rebuild`] has been called.
    pub fn layout_metrics(&self) -> LayoutMetrics {
        self.assert_rebuilt("layout_metrics");

        let first_baseline = self.layout.get(0).unwrap().metrics().baseline;
        let size = Size::new(self.layout.width().into(), self.layout.height().into());
        LayoutMetrics {
            size,
            first_baseline,
            trailing_whitespace_width: self.layout.full_width(),
        }
    }

    /// For a given `Point` (relative to this object's origin), returns index
    /// into the underlying text of the nearest grapheme boundary.
    ///
    /// This is not meaningful until [`Self::rebuild`] has been called.
    pub fn cursor_for_point(&self, point: Point) -> Cursor {
        self.assert_rebuilt("text_position_for_point");

        // TODO: This is a mostly good first pass, but doesn't handle cursor positions in
        // grapheme clusters within a parley cluster.
        // We can also try
        Cursor::from_point(&self.layout, point.x as f32, point.y as f32)
    }

    /// Given the utf-8 position of a character boundary in the underlying text,
    /// return the `Point` (relative to this object's origin) representing the
    /// boundary of the containing grapheme.
    ///
    /// # Panics
    ///
    /// Panics if `text_pos` is not a character boundary.
    ///
    /// This is not meaningful until [`Self::rebuild`] has been called.
    pub fn cursor_for_text_position(&self, text_pos: usize) -> Cursor {
        self.assert_rebuilt("cursor_for_text_position");

        // TODO: As a reminder, `is_leading` is not very useful to us; we don't know this ahead of time
        // We're going to need to do quite a bit of remedial work on these
        // e.g. to handle a inside a ligature made of multiple (unicode) grapheme clusters
        // https://raphlinus.github.io/text/2020/10/26/text-layout.html#shaping-cluster
        // But we're choosing to defer this work
        // This also needs to handle affinity.
        Cursor::from_position(&self.layout, text_pos, true)
    }

    /// Given the utf-8 position of a character boundary in the underlying text,
    /// return the `Point` (relative to this object's origin) representing the
    /// boundary of the containing grapheme.
    ///
    /// # Panics
    ///
    /// Panics if `text_pos` is not a character boundary.
    ///
    /// This is not meaningful until [`Self::rebuild`] has been called.
    pub fn point_for_text_position(&self, text_pos: usize) -> Point {
        let cursor = self.cursor_for_text_position(text_pos);
        Point::new(
            cursor.advance as f64,
            (cursor.baseline + cursor.offset) as f64,
        )
    }

    // TODO: needed for text selection
    // /// Given a utf-8 range in the underlying text, return a `Vec` of `Rect`s
    // /// representing the nominal bounding boxes of the text in that range.
    // ///
    // /// # Panics
    // ///
    // /// Panics if the range start or end is not a character boundary.
    // pub fn rects_for_range(&self, range: Range<usize>) -> Vec<Rect> {
    //     self.layout.rects_for_range(range)
    // }

    /// Given the utf-8 position of a character boundary in the underlying text,
    /// return a `Line` suitable for drawing a vertical cursor at that boundary.
    ///
    /// This is not meaningful until [`Self::rebuild`] has been called.
    // TODO: This is too simplistic. See https://raphlinus.github.io/text/2020/10/26/text-layout.html#shaping-cluster
    // for example. This would break in a `fi` ligature
    pub fn cursor_line_for_text_position(&self, text_pos: usize) -> Line {
        let from_position = self.cursor_for_text_position(text_pos);

        let line = from_position.path.line(&self.layout).unwrap();
        let line_metrics = line.metrics();

        let baseline = line_metrics.baseline + line_metrics.descent;
        let p1 = (from_position.offset as f64, baseline as f64);
        let p2 = (
            from_position.offset as f64,
            (baseline - line_metrics.size()) as f64,
        );
        Line::new(p1, p2)
    }

    /// Returns the [`Link`] at the provided point (relative to the layout's origin) if one exists.
    ///
    /// This can be used both for hit-testing (deciding whether to change the mouse cursor,
    /// or performing some other action when hovering) as well as for retrieving a [`Link`]
    /// on click.
    ///
    /// [`Link`]: super::attribute::Link
    pub fn link_for_pos(&self, pos: Point) -> Option<&Link> {
        let (_, i) = self
            .links
            .iter()
            .rfind(|(hit_box, _)| hit_box.contains(pos))?;

        self.text.links().get(*i)
    }

    /// Rebuild the inner layout as needed.
    ///
    /// This `TextLayout` object manages a lower-level layout object that may
    /// need to be rebuilt in response to changes to the text or attributes
    /// like the font.
    ///
    /// This method should be called whenever any of these things may have changed.
    /// A simple way to ensure this is correct is to always call this method
    /// as part of your widget's [`layout`][crate::Widget::layout] method.
    pub fn rebuild(
        &mut self,
        font_ctx: &mut FontContext,
        layout_ctx: &mut LayoutContext<TextBrush>,
    ) {
        self.rebuild_with_attributes(font_ctx, layout_ctx, |builder| builder);
    }

    /// Rebuild the inner layout as needed, adding attributes to the underlying layout.
    ///
    /// See [`Self::rebuild`] for more information
    pub fn rebuild_with_attributes(
        &mut self,
        font_ctx: &mut FontContext,
        layout_ctx: &mut LayoutContext<TextBrush>,
        attributes: impl for<'b> FnOnce(
            RangedBuilder<'b, TextBrush, &'b str>,
        ) -> RangedBuilder<'b, TextBrush, &'b str>,
    ) {
        if self.needs_layout {
            self.needs_layout = false;

            let mut builder = layout_ctx.ranged_builder(font_ctx, self.text.as_str(), self.scale);
            builder.push_default(&StyleProperty::Brush(self.brush.clone()));
            builder.push_default(&StyleProperty::FontSize(self.text_size));
            builder.push_default(&StyleProperty::FontStack(self.font));
            builder.push_default(&StyleProperty::FontWeight(self.weight));
            builder.push_default(&StyleProperty::FontStyle(self.style));
            // For more advanced features (e.g. variable font axes), these can be set in add_attributes

            let builder = self.text.add_attributes(builder);
            let mut builder = attributes(builder);
            builder.build_into(&mut self.layout);

            self.needs_line_breaks = true;
        }
        if self.needs_line_breaks {
            self.needs_line_breaks = false;
            self.layout
                .break_all_lines(self.max_advance, self.alignment);

            // TODO:
            // self.links = text
            //     .links()
            // ...
        }
    }

    /// Draw the layout at the provided `Point`.
    ///
    /// The origin of the layout is the top-left corner.
    ///
    /// You must call [`Self::rebuild`] at some point before you first
    /// call this method.
    pub fn draw(&mut self, scene: &mut Scene, point: impl Into<Point>) {
        self.assert_rebuilt("draw");
        // TODO: This translation doesn't seem great
        let p: Point = point.into();
        crate::text_helpers::render_text(
            scene,
            &mut self.scratch_scene,
            Affine::translate((p.x, p.y)),
            &self.layout,
        );
    }
}

impl<T: TextStorage> std::fmt::Debug for TextLayout<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        f.debug_struct("TextLayout")
            .field("text", &self.text.as_str().len())
            .field("scale", &self.scale)
            .field("brush", &self.brush)
            .field("font", &self.font)
            .field("text_size", &self.text_size)
            .field("weight", &self.weight)
            .field("style", &self.style)
            .field("alignment", &self.alignment)
            .field("wrap_width", &self.max_advance)
            .field("outdated?", &self.needs_rebuild())
            .field("width", &self.layout.width())
            .field("height", &self.layout.height())
            .finish()
    }
}

impl<T: TextStorage + Default> Default for TextLayout<T> {
    fn default() -> Self {
        Self::new(Default::default(), crate::theme::TEXT_SIZE_NORMAL as f32)
    }
}
