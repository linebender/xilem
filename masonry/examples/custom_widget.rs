// Copyright 2019 the Xilem Authors and the Druid Authors
// SPDX-License-Identifier: Apache-2.0

//! An example of a custom drawing widget.
//! We draw an image, some text, a shape, and a curve.

// On Windows platform, don't show a console when opening the app.
#![windows_subsystem = "windows"]

use accesskit::Role;
use kurbo::Stroke;
use masonry::app_driver::{AppDriver, DriverCtx};
use masonry::kurbo::BezPath;
use masonry::widget::{FillStrat, RootWidget, WidgetRef};
use masonry::{
    AccessCtx, AccessEvent, Action, Affine, BoxConstraints, Color, EventCtx, LayoutCtx, LifeCycle,
    LifeCycleCtx, PaintCtx, Point, PointerEvent, Rect, Size, StatusChange, TextEvent, Widget,
    WidgetId,
};
use parley::layout::Alignment;
use parley::style::{FontFamily, FontStack, StyleProperty};
use smallvec::SmallVec;
use tracing::{trace_span, Span};
use vello::peniko::{Brush, Fill, Format, Image};
use vello::Scene;
use winit::window::Window;

struct Driver;

impl AppDriver for Driver {
    fn on_action(&mut self, _ctx: &mut DriverCtx<'_>, _widget_id: WidgetId, _action: Action) {}
}

struct CustomWidget(String);

// If this widget has any child widgets it should call its event, update and layout
// (and lifecycle) methods as well to make sure it works. Some things can be filtered,
// but a general rule is to just pass it through unless you really know you don't want it.
impl Widget for CustomWidget {
    fn on_pointer_event(&mut self, _ctx: &mut EventCtx, _event: &PointerEvent) {}

    fn on_text_event(&mut self, _ctx: &mut EventCtx, _event: &TextEvent) {}

    fn on_access_event(&mut self, _ctx: &mut EventCtx, _event: &AccessEvent) {}

    fn lifecycle(&mut self, _ctx: &mut LifeCycleCtx, _event: &LifeCycle) {}

    fn on_status_change(&mut self, _ctx: &mut LifeCycleCtx, _event: &StatusChange) {}

    fn layout(&mut self, _layout_ctx: &mut LayoutCtx, bc: &BoxConstraints) -> Size {
        // BoxConstraints are passed by the parent widget.
        // This method can return any Size within those constraints:
        // bc.constrain(my_size)
        //
        // To check if a dimension is infinite or not (e.g. scrolling):
        // bc.is_width_bounded() / bc.is_height_bounded()
        //
        // bx.max() returns the maximum size of the widget. Be careful
        // using this, since always make sure the widget is bounded.
        // If bx.max() is used in a scrolling widget things will probably
        // not work correctly.
        if bc.is_width_bounded() && bc.is_height_bounded() {
            bc.max()
        } else {
            let size = Size::new(100.0, 100.0);
            bc.constrain(size)
        }
    }

    // The paint method gets called last, after an event flow.
    // It goes event -> update -> layout -> paint, and each method can influence the next.
    // Basically, anything that changes the appearance of a widget causes a paint.
    fn paint(&mut self, ctx: &mut PaintCtx, scene: &mut Scene) {
        // Clear the whole widget with the color of your choice
        // (ctx.size() returns the size of the layout rect we're painting in)
        // Note: ctx also has a `clear` method, but that clears the whole context,
        // and we only want to clear this widget's area.
        let size = ctx.size();
        let rect = size.to_rect();
        scene.fill(Fill::NonZero, Affine::IDENTITY, Color::WHITE, None, &rect);

        // Create an arbitrary bezier path
        let mut path = BezPath::new();
        path.move_to(Point::ORIGIN);
        path.quad_to((60.0, 120.0), (size.width, size.height));
        // Create a color
        let stroke_color = Color::rgb8(0, 128, 0);
        // Stroke the path with thickness 5.0
        scene.stroke(
            &Stroke::new(5.0),
            Affine::IDENTITY,
            stroke_color,
            None,
            &path,
        );

        // Rectangles: the path for practical people
        let rect = Rect::from_origin_size((10.0, 10.0), (100.0, 100.0));
        // Note the Color:rgba8 which includes an alpha channel (7F in this case)
        let fill_color = Color::rgba8(0x00, 0x00, 0x00, 0x7F);
        scene.fill(Fill::NonZero, Affine::IDENTITY, fill_color, None, &rect);

        // To render text, we first create a LayoutBuilder and set the text properties.
        let mut lcx = parley::LayoutContext::new();
        let mut text_layout_builder = lcx.ranged_builder(ctx.text_contexts().0, &self.0, 1.0);

        text_layout_builder.push_default(&StyleProperty::FontStack(FontStack::Single(
            FontFamily::Generic(parley::style::GenericFamily::Serif),
        )));
        text_layout_builder.push_default(&StyleProperty::FontSize(24.0));
        text_layout_builder.push_default(&StyleProperty::Brush(Brush::Solid(fill_color).into()));

        let mut text_layout = text_layout_builder.build();
        text_layout.break_all_lines(None, Alignment::Start);

        let mut scratch_scene = Scene::new();
        // We can pass a transform matrix to rotate the text we render
        masonry::text_helpers::render_text(
            scene,
            &mut scratch_scene,
            Affine::rotate(std::f64::consts::FRAC_PI_4).then_translate((80.0, 40.0).into()),
            &text_layout,
        );

        // Let's burn some CPU to make a (partially transparent) image buffer
        let image_data = make_image_data(256, 256);
        let image_data = Image::new(image_data.into(), Format::Rgba8, 256, 256);
        let transform = FillStrat::Fill.affine_to_fill(ctx.size(), size);
        scene.draw_image(&image_data, transform);
    }

    fn accessibility_role(&self) -> Role {
        Role::Window
    }

    fn accessibility(&mut self, ctx: &mut AccessCtx) {
        let text = &self.0;
        ctx.current_node().set_name(
            format!("This is a demo of the Masonry Widget trait. Masonry has accessibility tree support. The demo shows colored shapes with the text '{text}'."),
        );
    }

    fn children(&self) -> SmallVec<[WidgetRef<'_, dyn Widget>; 16]> {
        SmallVec::new()
    }

    fn make_trace_span(&self) -> Span {
        trace_span!("CustomWidget")
    }
}

pub fn main() {
    let my_string = "Masonry + Vello".to_string();
    let window_attributes = Window::default_attributes().with_title("Fancy colors");

    masonry::event_loop_runner::run(
        masonry::event_loop_runner::EventLoop::with_user_event(),
        window_attributes,
        RootWidget::new(CustomWidget(my_string)),
        Driver,
    )
    .unwrap();
}

fn make_image_data(width: usize, height: usize) -> Vec<u8> {
    let mut result = vec![0; width * height * 4];
    for y in 0..height {
        for x in 0..width {
            let ix = (y * width + x) * 4;
            result[ix] = x as u8;
            result[ix + 1] = y as u8;
            result[ix + 2] = !(x as u8);
            result[ix + 3] = 127;
        }
    }
    result
}
