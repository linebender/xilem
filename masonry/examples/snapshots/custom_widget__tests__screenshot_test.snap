---
source: masonry/examples/custom_widget.rs
assertion_line: 195
expression: harness.root_widget()
---
CustomWidget
